SHARD-2463: Handle receipt and original tx thorugh storeReceiptData

S0naliThakur opened the pull request.

S0naliThakur committed:
Refactored checkpoint logic: `storeReceiptData` now handles originalTx checkpointing. Removed deprecated originalTx saving logic from checkpoint mode.

S0naliThakur committed:
Added import for `storeReceiptData` in `ReceiptData.ts` to support updated checkpointing logic.

aniketdivekar commented on src/Data/Collector.ts:
```diff
@@ -728,7 +728,8 @@ export const storeReceiptData = async (
   verifyData = false,
-  saveOnlyGossipData = false
+  saveOnlyGossipData = false,
+  checkpoint: boolean = true
```
this should be by default false right? only from checkpointing code we should pass it as true.

S0naliThakur replied:
This ensures consistency when creating new checkpoints as data is inserted into the database. By default, the database methods set this flag to true. Only the checkpoint system sets it to false, which is why it's true by default here.

aniketdivekar commented on src/Data/Collector.ts:
```diff
@@ -1297,7 +1298,8 @@ export const storeAccountData = async (restoreData: StoreAccountParam = {}): Pro
   senderInfo = '',
-  saveOnlyGossipData = false
+  saveOnlyGossipData = false,
+  checkpoint: boolean = true
```
we don't need this change as no where are we invoking storeOriginalTxData - its technically deprecated

S0naliThakur replied:
You are right, it's deprecated and not used, but we made the change for consistency, just to keep things aligned across the board. That said, I can remove it if you prefer.

aniketdivekar replied:
Please go ahead and revert the changes. If need be, add a note on top of the method that its deprecated.

S0naliThakur replied:
done

S0naliThakur committed:
fix storeReceiptdata checkpoint flow

S0naliThakur committed:
Deprecate `storeOriginalTxData` method and remove checkpoint parameter

devendra-shardeum commented on src/Data/Collector.ts:
```diff
@@ -765,8 +766,9 @@ export const storeReceiptData = async (
       if (
-        (processedReceiptsMap.has(txId) && processedReceiptsMap.get(txId) === timestamp) ||
-        (receiptsInValidationMap.has(txId) && receiptsInValidationMap.get(txId) === timestamp)
+        checkpoint &&
```
please do not add the checkpoint flag here.. since the invalidation maps needs to be checked regardless otherwise it'll skip and do the processing of receipts which has already been seen in past

devendra-shardeum replied:
also this way we protect our archiver from potential dos as well.

S0naliThakur replied:
The checks will always happen in case of normal flow. But since checkpoint works for recent transactions only, this check will not let us save the majority agreed upon receipt that's why checkpoint flag is needed here.

devendra-shardeum commented on src/Data/Collector.ts:
```diff
@@ -788,7 +790,7 @@ export const storeReceiptData = async (
         // only consider this for EVM txns and Non Global Internal Txns
         const result = await checkIfValidOverwrite(receipt, txId)
-        if (!result) {
+        if (!result && checkpoint) {
```
this should always be checked regardless of checkpoint enabled or not.. why did we add this flag here ?

S0naliThakur replied:
As addressed in the slack conversation, this is needed to save the newly agreed upon success receipt.

devendra-shardeum commented on src/checkpoint/ReceiptData.ts:
```diff
@@ -115,7 +116,7 @@ async function updateData(data: CheckpointData<ReceiptType>): Promise<void> {
     const receipt = data.d
     // Avg entry size is about 41000 bytes
-    await insertReceipt(receipt, false)
+    await storeReceiptData([receipt], 'checkpoint', false, false, false)
```
Please update all calls to storeReceiptData to explicitly pass the flags as arguments. This will improve code readability and make it clearer what each call is intending to do.

![image](https://github.com/user-attachments/assets/9c0040de-6f4b-4795-acb8-11ae76e474cf)

S0naliThakur replied:
Added

S0naliThakur committed:
Refactor `storeReceiptData` calls to include additional parameters

S0naliThakur committed:
Add receipt selection logic to improve checkpoint processing

- Introduced `selectBestSuccessReceipt` function to determine the best receipt based on consensus rules.
- Updated `CheckpointBucket` to utilize the new receipt selection logic when handling multiple receipts.
- Added a utility function to strip non-consensus fields from receipts for accurate hashing.
- Enhanced `ReceiptCheckpointData` to use the stripped receipt for hash calculations.
- Included comprehensive unit tests for the new receipt selection functionality.

urnotsam commented:
Please fix failing tests

S0naliThakur committed:
mass format

S0naliThakur committed:
fix unit tests in `ReceiptData.test.ts`
